`ConstantFoldingRewriter` aborts module load when Jython `PythonInterpreter` init fails (`sys.importlib is null`); silently empties entrypoints

## Summary

`PropagationCallGraphBuilder.makeCallGraph` throws `IllegalStateException: Could not create a entrypoint callsites:` (with an **empty** `Warnings.asString()` tail, which strongly implies `options.getEntrypoints()` returned an empty iterable rather than that every entrypoint's `addCall` failed) on three Hybridize tests that exercise heavy Keras/Dataset usage. Affects 3 `HybridizeFunctionRefactoringTest` cases under Ariadne 0.41 (`5467c89a`):

- `testAutoEncoder` — `tf.initializers.RandomNormal()`, `tf.GradientTape()`, `optimizer.apply_gradients(...)`, `mnist.load_data()`.
- `testTensorFlowEagerExecution` — `keras.Model` subclass with `Conv2D`/`Flatten`/`Dense`, `mnist.load_data()`, two `@tf.function`-decorated training functions.
- `testDatasetIteration4` — `from tensorflow.keras.preprocessing.image import ImageDataGenerator` + a `@tf.function distributed_train_step`.

All three were green on Hybridize `main` (`bb0ab35`) with the previous summary set, so the regression is in the new Ariadne (or in the converged tensorflow.xml's interaction with it), not in client code.

## Stack

```
java.lang.IllegalStateException: Could not create a entrypoint callsites: 
    at com.ibm.wala.ipa.callgraph.propagation.PropagationCallGraphBuilder.makeCallGraph(PropagationCallGraphBuilder.java:238)
    at edu.cuny.hunter.hybridize.core.refactorings.HybridizeFunctionRefactoringProcessor.computeCallGraph(...:472)
    at ...checkFunctions(...:297)
    at ...checkFinalConditions(...:258)
    at ...HybridizeFunctionRefactoringTest.testAutoEncoder(...:6837)
```

The trailing-empty `Warnings.asString()` is the same in all three, suggesting `options.getEntrypoints()` itself returned empty — i.e., the Python entrypoint scanner discovered nothing — rather than the resolution loop running but rejecting every candidate.

## Common-shape conjecture (needs verification with `DEBUG_ENTRYPOINTS=true`)

The three inputs share aggressive Keras/Dataset surface that bottoms out into `from tensorflow.keras.{datasets,preprocessing.image} import …`. If a regression in the front-end or in `tensorflow.xml` is causing one of these imports to fail to resolve, the whole module load may be aborted and the entrypoint scanner left with no module to walk.

## Reproduction

```bash
git clone https://github.com/ponder-lab/Hybridize-Functions-Refactoring
cd Hybridize-Functions-Refactoring && git checkout fac28ee   # tensorflow.xml convergence
./mvnw -s .travis.settings.xml verify -pl edu.cuny.hunter.hybridize.tests \
       -Dtest=HybridizeFunctionRefactoringTest#testAutoEncoder
```

## Asks

- Confirm whether this is a known/expected regression of the dataset/keras frontend changes that landed alongside `5467c89a`.
- If not, what's the cleanest way to surface *which* entrypoint discovery step gave up — flag, log level, or extra warning text would all unblock client-side triage.

## Pinpointed root cause

The NPE originates inside `Python3Interpreter.getInterp()` when it constructs a fresh `PythonInterpreter`:

```java
public static PythonInterpreter getInterp() {
  if (interp == null) {
    PySystemState.initialize();
    interp = new PythonInterpreter();   // ← throws NPE during site import
  }
  return interp;
}
```

Stderr shows the underlying cause is a missing resource:

```
java.io.FileNotFoundException: src/resources/frozen_importlib/_frozen_importlib.class (No such file or directory)
Caused by: java.lang.NullPointerException: Cannot invoke "...PyObject.invoke(...)" because "sys.importlib" is null
```

Jython looks for `_frozen_importlib.class` to bootstrap importlib; when it can't find it, `sys.importlib` is left null and `Py.importSiteIfSelected` NPEs. The exception propagates up through `Python3Loader$4$1.eval(Python3Loader.java:112)` — which only catches `PySyntaxError`, so the NPE escapes. It then bubbles through every recursive `ConstantFoldingRewriter.copyNodes` frame on the stack, aborting the rewrite entirely. The class loader drops the module, the CHA has no `.py` entry, and `makeDefaultEntrypoints` returns empty.

## Two-place narrow fix

Either of these would be sufficient (both is best):

**1. Broaden the catch in `Python3Loader.java:113`** to swallow non-`PySyntaxError` failures from the embedded interpreter — those represent infrastructure problems (missing Jython resource, JVM env issue) rather than meaningful "this expression can't be folded" signal:

```java
try {
  x = Python3Interpreter.getInterp().eval(unicode);
} catch (PySyntaxError e) {
  logger.log(WARNING, e, () -> "Syntax error in expression: " + unicode);
  return null;
} catch (RuntimeException | Error e) {       // ← add this
  logger.log(WARNING, e, () -> "Const-fold disabled (interp init failed): " + unicode);
  return null;
}
```

**2. Cache the failure in `Python3Interpreter.getInterp()`** so each test doesn't retry the failing init dozens of times:

```java
private static volatile boolean initFailed = false;

public static PythonInterpreter getInterp() {
  if (initFailed) return null;
  if (interp == null) {
    try {
      PySystemState.initialize();
      interp = new PythonInterpreter();
    } catch (Throwable t) {
      initFailed = true;
      throw t;
    }
  }
  return interp;
}
```

(A `null` return then needs the same handling at every call site as a regular const-fold miss.)

## Why this is environment-sensitive

The fact that `testTensorFlowGanTutorial2` shows the same NPE in stderr **but still passes**, while `testAutoEncoder`/`testTensorFlowEagerExecution`/`testDatasetIteration4` fail, suggests the NPE is recoverable in some module-load paths but fatal in others — likely depending on whether the const-fold rewrite is invoked at the top of a tree (so the whole module dies) or only on a sub-tree (so the module survives with a missing rewrite).

`Python3Interpreter.getInterp()` running fine in Ariadne's own `mvn verify` (608 tests, 0 errors related to Jython) but failing in Tycho's surefire-osgi-booter is also consistent — the OSGi bundle classloader probably can't see `src/resources/frozen_importlib/_frozen_importlib.class` from the jython3 resources jar the way Maven surefire can. That's a bundle-classpath issue, not a code bug, but the recovery patch in (1) makes Ariadne resilient to it regardless.

## Reproducer (Hybridize side)

```bash
cd Hybridize-Functions-Refactoring && git checkout fac28ee
./mvnw -s .travis.settings.xml verify \
       -pl edu.cuny.hunter.hybridize.tests \
       -Dtest=HybridizeFunctionRefactoringTest#testAutoEncoder
```

Surefire stderr will contain the `_frozen_importlib.class` `FileNotFoundException` followed by the `sys.importlib` NPE; the test reports `IllegalStateException: Could not create a entrypoint callsites:` (with empty `Warnings`) at `PropagationCallGraphBuilder.makeCallGraph:238`.

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
@@ -103,18 +103,20 @@ public class Python3Loader extends PythonLoader {
               @Override
               protected Object eval(CAstOperator op, Object lhs, Object rhs) {
                 String s = lhs + " " + op.getValue() + " " + rhs;
+
+                PythonInterpreter ip = Python3Interpreter.getInterp();
+                if (ip == null) {
+                  // Jython init failed (memoized in Python3Interpreter). Skip constant folding
+                  // for this expression; analysis remains correct, just less precise. Don't log
+                  // an "Evaluating:" entry — nothing is actually evaluated, and the underlying
+                  // init failure was already announced from getInterp().
+                  return null;
+                }
                 logger.info(() -> "Evaluating: " + s);
 
                 // Use the Python interpreter to evaluate the expression.
                 PyUnicode unicode = new PyUnicode(s);
                 PyObject x;
-
-                PythonInterpreter ip = Python3Interpreter.getInterp();
-                if (ip == null) {
-                  // Jython init failed (memoized in Python3Interpreter). Skip constant folding
-                  // for this expression; analysis remains correct, just less precise.
-                  return null;
-                }
 
                 try {
                   x = ip.eval(unicode);

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
@@ -108,8 +108,15 @@ public class Python3Loader extends PythonLoader {
                 PyUnicode unicode = new PyUnicode(s);
                 PyObject x;
 
+                org.python.util.PythonInterpreter ip = Python3Interpreter.getInterp();
+                if (ip == null) {
+                  // Jython init failed (memoized in Python3Interpreter). Skip constant folding
+                  // for this expression; analysis remains correct, just less precise.
+                  return null;
+                }
+
                 try {
-                  x = Python3Interpreter.getInterp().eval(unicode);
+                  x = ip.eval(unicode);
                 } catch (PySyntaxError e) {
                   // Handle syntax errors gracefully.
                   logger.log(WARNING, e, () -> "Syntax error in expression: " + unicode);

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/loader/Python3Loader.java
@@ -38,6 +38,7 @@ import java.util.logging.Logger;
 import org.python.core.PyObject;
 import org.python.core.PySyntaxError;
 import org.python.core.PyUnicode;
+import org.python.util.PythonInterpreter;
 
 public class Python3Loader extends PythonLoader {
 
@@ -108,7 +109,7 @@ public class Python3Loader extends PythonLoader {
                 PyUnicode unicode = new PyUnicode(s);
                 PyObject x;
 
-                org.python.util.PythonInterpreter ip = Python3Interpreter.getInterp();
+                PythonInterpreter ip = Python3Interpreter.getInterp();
                 if (ip == null) {
                   // Jython init failed (memoized in Python3Interpreter). Skip constant folding
                   // for this expression; analysis remains correct, just less precise.

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -30,10 +30,10 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
    * failure. The first failure is already announced by the catch block in {@link #getInterp()};
    * subsequent calls log at FINE level only.
    *
-   * <p>Uses {@link AtomicBoolean#compareAndSet} so the check-and-set is atomic — under concurrent
-   * call graph construction, multiple threads can race into the {@code if (ip == null)} branch
-   * simultaneously, and a non-atomic {@code volatile boolean} flag would let several of them all
-   * pass the check before any sets it, defeating the "log once" intent.
+   * @implNote Uses {@link AtomicBoolean#compareAndSet} so the check-and-set is atomic. Under
+   *     concurrent call graph construction, multiple threads can race into the {@code if (ip ==
+   *     null)} branch simultaneously; a non-atomic {@code volatile boolean} flag would let several
+   *     of them all pass the check before any sets it, defeating the "log once" intent.
    */
   private static final AtomicBoolean unavailableWarned = new AtomicBoolean(false);
 

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -53,10 +53,19 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
   public Integer evalAsInteger(String expr) {
     PythonInterpreter ip = getInterp();
     if (ip == null) {
-      throw new IllegalStateException(
-          "Jython interpreter unavailable (init failed earlier); cannot evaluate expression: "
-              + expr
-              + ".");
+      // Return {@code null} (the same "cannot evaluate" signal used elsewhere in this method's
+      // contract) rather than throwing, so callers like
+      // {@code com.ibm.wala.cast.python.util.PythonInterpreter#interpretAsInt} — which expect a
+      // nullable {@link Integer} and don't catch checked or runtime exceptions — degrade
+      // gracefully in the same OSGi-classloader environments that triggered the {@link
+      // #getInterp()} init failure in the first place.
+      LOGGER.log(
+          Level.WARNING,
+          () ->
+              "Jython interpreter unavailable (init failed earlier); cannot evaluate expression: "
+                  + expr
+                  + ". Returning null.");
+      return null;
     }
     try {
       PyObject val = ip.eval(expr);

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -22,6 +22,15 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
    */
   private static volatile boolean initFailed = false;
 
+  /**
+   * Memoizes whether the "interpreter unavailable" warning has already been emitted from {@link
+   * #evalAsInteger(String)}. Without this, callers like {@code interpretAsInt} (invoked many times
+   * during shape inference) would flood logs with one WARNING per call after the first init
+   * failure. The first failure is already announced by the catch block in {@link #getInterp()};
+   * subsequent calls log at FINE level only.
+   */
+  private static volatile boolean unavailableWarned = false;
+
   public static synchronized PythonInterpreter getInterp() {
     if (initFailed) return null;
     if (interp == null) {
@@ -43,7 +52,9 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
             Level.WARNING,
             t,
             () ->
-                "Jython interpreter init failed; constant folding will be disabled for this run.");
+                "Jython interpreter init failed; all interpreter-based evaluation will be disabled"
+                    + " for this run (constant folding in Python3Loader, shape-argument"
+                    + " evaluation via interpretAsInt/evalAsInteger, etc.).");
         return null;
       }
     }
@@ -59,12 +70,21 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
       // nullable {@link Integer} and don't catch checked or runtime exceptions — degrade
       // gracefully in the same OSGi-classloader environments that triggered the {@link
       // #getInterp()} init failure in the first place.
-      LOGGER.log(
-          Level.WARNING,
-          () ->
-              "Jython interpreter unavailable (init failed earlier); cannot evaluate expression: "
-                  + expr
-                  + ". Returning null.");
+      //
+      // Log the first such call at WARNING (so operators see that some shape inference is being
+      // skipped because of the earlier init failure); subsequent calls log at FINE only, since
+      // the underlying init failure has already been announced from {@link #getInterp()}.
+      if (!unavailableWarned) {
+        unavailableWarned = true;
+        LOGGER.log(
+            Level.WARNING,
+            () ->
+                "Jython interpreter unavailable (init failed earlier); evalAsInteger will return"
+                    + " null for this and any subsequent calls. First skipped expression: "
+                    + expr);
+      } else {
+        LOGGER.log(Level.FINE, () -> "evalAsInteger returning null (interp unavailable): " + expr);
+      }
       return null;
     }
     try {

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -22,18 +22,22 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
    */
   private static volatile boolean initFailed = false;
 
-  public static PythonInterpreter getInterp() {
+  public static synchronized PythonInterpreter getInterp() {
     if (initFailed) return null;
     if (interp == null) {
       try {
         PySystemState.initialize();
         interp = new PythonInterpreter();
-      } catch (Throwable t) {
+      } catch (Exception t) {
         // Jython init can fail when bootstrap resources (e.g., the embedded
         // _frozen_importlib bytecode used by org.python.core.imp) aren't reachable from the
         // current classloader/working directory. This is environment-dependent (e.g., happens
         // under Tycho-OSGi but not under plain Maven-surefire). Treat as a recoverable failure:
         // log once, memoize, and let callers degrade gracefully.
+        //
+        // We catch {@link Exception} (not {@link Throwable}) so that {@link Error} types
+        // (OOM, stack overflow, linkage errors) keep propagating to the caller — those signal
+        // serious VM problems we don't want to silently swallow and continue past.
         initFailed = true;
         LOGGER.log(
             Level.WARNING,
@@ -47,8 +51,15 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
   }
 
   public Integer evalAsInteger(String expr) {
+    PythonInterpreter ip = getInterp();
+    if (ip == null) {
+      throw new IllegalStateException(
+          "Jython interpreter unavailable (init failed earlier); cannot evaluate expression: "
+              + expr
+              + ".");
+    }
     try {
-      PyObject val = getInterp().eval(expr);
+      PyObject val = ip.eval(expr);
       if (val.isInteger()) {
         return val.asInt();
       } else

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -1,5 +1,6 @@
 package com.ibm.wala.cast.python.util;
 
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import org.python.core.PyException;
@@ -28,8 +29,13 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
    * during shape inference) would flood logs with one WARNING per call after the first init
    * failure. The first failure is already announced by the catch block in {@link #getInterp()};
    * subsequent calls log at FINE level only.
+   *
+   * <p>Uses {@link AtomicBoolean#compareAndSet} so the check-and-set is atomic — under concurrent
+   * call graph construction, multiple threads can race into the {@code if (ip == null)} branch
+   * simultaneously, and a non-atomic {@code volatile boolean} flag would let several of them all
+   * pass the check before any sets it, defeating the "log once" intent.
    */
-  private static volatile boolean unavailableWarned = false;
+  private static final AtomicBoolean unavailableWarned = new AtomicBoolean(false);
 
   public static synchronized PythonInterpreter getInterp() {
     if (initFailed) return null;
@@ -74,8 +80,7 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
       // Log the first such call at WARNING (so operators see that some shape inference is being
       // skipped because of the earlier init failure); subsequent calls log at FINE only, since
       // the underlying init failure has already been announced from {@link #getInterp()}.
-      if (!unavailableWarned) {
-        unavailableWarned = true;
+      if (unavailableWarned.compareAndSet(false, true)) {
         LOGGER.log(
             Level.WARNING,
             () ->

--- a/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
+++ b/com.ibm.wala.cast.python.jython3/source/com/ibm/wala/cast/python/util/Python3Interpreter.java
@@ -13,10 +13,35 @@ public class Python3Interpreter extends com.ibm.wala.cast.python.util.PythonInte
 
   private static PythonInterpreter interp;
 
+  /**
+   * Memoizes a failed Jython init so subsequent {@link #getInterp()} calls return {@code null}
+   * cheaply instead of re-running {@code new PythonInterpreter()} (which can be expensive when it
+   * fails — site-import walks the Jython resource path on every attempt). When a single failure has
+   * occurred, callers receive {@code null} and can degrade their behavior (e.g., {@link
+   * com.ibm.wala.cast.python.loader.Python3Loader} skips constant folding).
+   */
+  private static volatile boolean initFailed = false;
+
   public static PythonInterpreter getInterp() {
+    if (initFailed) return null;
     if (interp == null) {
-      PySystemState.initialize();
-      interp = new PythonInterpreter();
+      try {
+        PySystemState.initialize();
+        interp = new PythonInterpreter();
+      } catch (Throwable t) {
+        // Jython init can fail when bootstrap resources (e.g., the embedded
+        // _frozen_importlib bytecode used by org.python.core.imp) aren't reachable from the
+        // current classloader/working directory. This is environment-dependent (e.g., happens
+        // under Tycho-OSGi but not under plain Maven-surefire). Treat as a recoverable failure:
+        // log once, memoize, and let callers degrade gracefully.
+        initFailed = true;
+        LOGGER.log(
+            Level.WARNING,
+            t,
+            () ->
+                "Jython interpreter init failed; constant folding will be disabled for this run.");
+        return null;
+      }
     }
     return interp;
   }

